Group create function

robot269006 opened the pull request:
# WHAT
Enable function to create groups

# WHY
Allow users to make groups as they wish, and also to belong to groups

robot269006 committed:
Add Create function to groups

1. Add “create” method
2. Allow “create” parameters

robot269006 committed:
Merge branch 'group-model-controller' into group-create-function

robot269006 committed:
Revise group creation function

1. Delete :body in params.require(:group).permit(:name), since :body is
no longer required.
2. rails s => “save” button pushed => doesn’t work.
3. Add pry gem, binding.pry
4. Implement binding.pry (in groups_controller, in create method), to
see what parameters are being sent from new.haml
5. Parameters that were not set in Groups table (user_id => 22,
search-field) temporarily commented out.

*Trying to “save” new group leads to “No routes match”
 -> Looked at the new.haml, and found that there was an action set as
default (:action => “/chat_groups”), which doesn’t exist. So deleted
this whole line, replacing with form_for with @group instance.

robot269006 committed:
Create mid-table and add columns on user and group tables

In order to set other parameters sent from new.haml:

1. Create mid-table group_user (rails g model group_user)
2. Edit migration, and rake db:migrate
3. Add columns to user and group table by rails g migration
AddColumnToUser and AddColumnToGroup
4. rake db:migrate

robot269006 committed:
Merge branch 'group-model-controller' into group-create-function

# Conflicts:
#	app/views/devise/edit.html.erb
#	app/views/devise/edit.html.haml
#	app/views/user/edit.html.haml

robot269006 committed:
Merge branch 'master' into group-create-function

robot269006 committed:
Set association

1. rails g migration Group_user
2. Set association (has_many :groups, users)
3. Set accepts_nested_attributes_for

robot269006 committed:
Revise group_user table

rails g migration RemoveColFromGroupuser
-> Correct mistake of making “group_id_id” trying to add reference and
foreign_key

robot269006 committed:
Set redirection

1. Set redirection on groups_controller

FumiyaMori commented on .gitignore:
```diff
@@ -11,3 +11,5 @@
 !/log/.keep
 /tmp
+
+app/views/groups/new.html.haml
```
このファイルは必要なのでgitの管理から無視しないほうが良いのですが、何か理由はありますか？

robot269006 replied:
このPRをレビューいただいている間に別のブランチからアップデートした際にConflictが表示されたので、一度その通知をIgnoreし、別ブランチでいただいた修正指摘を直接反映させました。なので、その時のIgnore指示が残ったままなのだと思います。単純にファイルからこの記述を取ればいいのでしょうか？

FumiyaMori replied:
この行をファイルから削除してコミットすれば大丈夫だと思います

robot269006 replied:
承知しました、ご確認ありがとうございます。

FumiyaMori commented on app/controllers/groups_controller.rb:
```diff
@@ -1,9 +1,20 @@
   end
 
+  def create
+    @group = Group.create(create_params)
```
createを使うのではなくnewとsaveに分け、保存に成功した場合失敗した場合の処理をそれぞれ書きましょう

robot269006 replied:
Ifでsaveとelseに分ける感じでしょうか？new自体は def new endの中に記述しようと思います。

FumiyaMori replied:
newアクションとは別にcreateアクション内でもGroup.newをする必要があります。
その上でsaveメソッドを呼び出したときの返り値を使い、条件分岐をしてみましょう

FumiyaMori commented on db/migrate/20170417110310_add_column_to_user.rb:
```diff
@@ -0,0 +1,5 @@
+class AddColumnToUser < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :group_user_id, :integer
```
正しくアソシエーションが組まれていればこのカラムは必要ないと思います。
もう一度確認をお願いします

robot269006 replied:
中間テーブルのgroup_userのidなのですが、UserとGroupそれぞれのモデルに「through: :中間テーブル」の記述があればこのカラムは特に必要ないという理解で合ってますでしょうか？

FumiyaMori replied:
その理解であっていると思います

robot269006 replied:
わかりました。では、groupとuserテーブルそれぞれで、group_user_idのカラムは削除します。

FumiyaMori commented on db/migrate/20170417110700_add_column_to_groups.rb:
```diff
@@ -0,0 +1,5 @@
+class AddColumnToGroups < ActiveRecord::Migration[5.0]
+  def change
+    add_column :groups, :group_user_id, :integer
```
こちらも同様に、正しくアソシエーションが組まれていればこのカラムは必要ないと思います。
もう一度確認をお願いします

FumiyaMori commented on .gitignore:
```diff
@@ -11,3 +11,5 @@
 !/log/.keep
 /tmp
+
+app/views/groups/new.html.haml
```
この行をファイルから削除してコミットすれば大丈夫だと思います

FumiyaMori commented on app/controllers/groups_controller.rb:
```diff
@@ -1,9 +1,20 @@
   end
 
+  def create
+    @group = Group.create(create_params)
```
newアクションとは別にcreateアクション内でもGroup.newをする必要があります。
その上でsaveメソッドを呼び出したときの返り値を使い、条件分岐をしてみましょう

FumiyaMori commented on db/migrate/20170417110310_add_column_to_user.rb:
```diff
@@ -0,0 +1,5 @@
+class AddColumnToUser < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :group_user_id, :integer
```
その理解であっていると思います

robot269006 committed:
Revise (gitignore and group_user_id column removal)

1. Delete the last line on gitignore to allow git check on
/groups/new.html.haml
2. Remove group_user_id column both from group table and user table

robot269006 committed:
Revision (new and save on create method)

1. Edit create method from Groups.create to Groups.new
2. Set if conditions -> to make flash messages appear

vip-take commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,27 @@
+.chat-group-form
+  %h1 新規チャットグループ
+  = form_for (@group) do |f|
```
括弧()は無くても大丈夫なはずです。

robot269006 replied:
承知しました。ご確認ありがとうございます！

vip-take commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,27 @@
+  = form_for (@group) do |f|
+    .chat-group-form__field.clearfix
+      .chat-group-form__field--left
+        %label.chat-group-form__label{for: "group[name]"} グループ名
```
form_forを使っているので、f.のような形に直しましょう！

robot269006 replied:
承知しました。ご確認ありがとうございます！

vip-take commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,27 @@
+      .chat-group-form__field--left
+        %label.chat-group-form__label{for: "group[name]"} グループ名
+      .chat-group-form__field--right
+        %input#chat_group_name.chat-group-form__input{name: "group[name]", placeholder: "グループ名を入力してください", type: "text"}/
```
form_forを使っているので、f.のような形に直しましょう！

robot269006 replied:
承知しました。ご確認ありがとうございます！

vip-take commented on app/views/groups/new.html.haml:
```diff
@@ -0,0 +1,27 @@
+    .chat-group-form__field.clearfix
+      .chat-group-form__field--left
+      .chat-group-form__field--right
+        %input.chat-group-form__action-btn{"data-disable-with": "Save", name: "commit", type: "submit", value: "Save"}/
```
こちらも、form_forを使っているので、f.のような形に直しましょう！

robot269006 replied:
承知しました。ご確認ありがとうございます！

vip-take commented on db/schema.rb:
```diff
@@ -10,20 +10,21 @@
-    t.datetime "created_at",                  null: false
-    t.datetime "updated_at",                  null: false
-    t.integer  "group_user_id"
+    t.text     "name",       limit: 65535, null: false
```
uniqueの制約もつけてあげましょう！

vip-take replied:
[参考]
この場合、型はtextではなく、できればstringの方が良いですね。
chatspaceでは不要ですが、検索効率を上げるためにindexを設定する際に、
一手間がかかります。

robot269006 replied:
承知しました、ご確認ありがとうございます！

robot269006 committed:
Revision (View file revision)

edit new.html.haml, where %input -> = f.label

robot269006 committed:
Revision (group table, add unique to name column)

1. rails g migration change_name_to_groups
2. change_column :groups, :name, :string, null: false
3. bundle exec rake db:migrate

robot269006 commented:
修正させていただきました。ビューの修正も行っているので、画像をアップします

robot269006 commented:
<img width="1226" alt="2017-04-19 17 16 06" src="https://cloud.githubusercontent.com/assets/24761206/25172001/aeb213f8-252a-11e7-8aa6-03107d07fb2f.png">
<img width="1195" alt="2017-04-19 18 03 00" src="https://cloud.githubusercontent.com/assets/24761206/25172000/aeab7b74-252a-11e7-8800-32b6978e083e.png">

kenchan0503 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,9 +1,26 @@
+    @group = Group.new(create_params)
+    if @group.save
+      flash[:notice]="チャットグループが作成されました"
+      redirect_to controller: :top, action: :index
```
redirect_toとflashメッセージはまとめて１行に定義しましょう

kenchan0503 replied:
controllerとactionオプションがもう少し完結に書けそうです。

kenchan0503 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,9 +1,26 @@
+      redirect_to controller: :top, action: :index
+    else
+      flash[:alert]="チャットグループの作成に失敗しました"
+      redirect_to controller: :groups, action: :new
```
こちらredirect_toよりもrenderの方が適切なのですがrenderとredirect_toについて調べてみてください。

kenchan0503 replied:
両者の違いについて気づいた点などをこちらに明示していただきますと幸いです。

robot269006 replied:
違い：renderは直接viewを呼び出し、redirect_toはパスを指定してそのパスが指定するviewを出す。redirect_toはrouteを経由してviewを表示するため、コントローラのアクション内で定義されているインスタンス変数をviewに渡すことができない

勝手な理解ですが、通常であればroute→controller→viewで表示するHTMLを指定し、そこに特定の情報（インスタンス変数）を渡して表示する所が、controller(redirect_to)→route→viewという順番になってしまうため、viewでインスタンス変数を使って何かを表示させようとしてもできない、という感じだと思います。

kenchan0503 replied:
その通りですね。

kenchan0503 replied:
ただ、saveが失敗した時にインスタンスの情報を保持したい(フォームの情報を残したい)という目的があるので基本的にsaveが失敗した時はrenderの方がよいです。ですのでcreateアクション内のインスタンスがrender: newの時にも対応できるようにコントローラー内にインスタンスを定義してあげましょう

robot269006 replied:
ありがとうございます。これはちょっとよくわかっていないのですが、redirect_to root_path(@group)というように修正するということでしょうか？（renderのときはインスタンスを渡せるけどredirect_toのときは渡せない、という理由で）

kenchan0503 replied:
いえ、saveがfalse時にはrender: new のようにコントローラーを経由せずにテンプレートだけ呼び出した方がよいということです。言葉足らずで申し訳ございません。。

robot269006 replied:
ご指摘ありがとうございます。修正させていただきました。

kenchan0503 commented on db/migrate/20170417102258_create_group_users.rb:
```diff
@@ -0,0 +1,9 @@
+class CreateGroupUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :group_users do |t|
+      t.integer     :user_id,   null: false
```
外部キーはpreference型で定義しましょう

kenchan0503 replied:
外部キー制約も合わせてお願いいたします

kenchan0503 replied:
外部キー制約はこちらですね。
foreign_key: true

参考
https://www.dbonline.jp/mysql/table/index11.html

robot269006 replied:
なるほど、SQLで直接設定しにいくんですね。そしてIndex貼ってないとforeign_keyの設定はできないんですね。。今後の参考にさせていただきます！

robot269006 replied:
ちなみにこの外部キー制約の設定は一旦問題なしでしょうか？

kenchan0503 replied:
foreign_key: true
などでいかがでしょうか？

robot269006 replied:
ちょっと要領を得ないのですが、schemaの最後の方に
  add_foreign_key "group_users", "groups", column: "groups_id"
  add_foreign_key "group_users", "users", column: "users_id"
と記載されてますがこれでは対応できてないのでしょうか？

robot269006 replied:
一度groups_userのテーブルをドロップしてから再度作り直してます。

robot269006 replied:
<img width="1276" alt="2017-04-19 21 11 38" src="https://cloud.githubusercontent.com/assets/24761206/25179475/09e6a7e2-2545-11e7-9b74-94d5d68a6739.png">

robot269006 replied:
修正の意図がよく分からないので、直接教えていただけないでしょうか？

robot269006 committed:
Revision (further edit on group_controller)

robot269006 committed:
Revision (reference and foreign_key on group_user)

1. Destroy group_user -> rails destroy model group_user
2. Drop table by rollback -> rails g migration
drop_table :group_users
3. bundle exec rake db:migrate
4. rails g model group_user
5. Edit migration file -> t.reference, foreign_key: true
6. bundle exec rake db:migrate
7. Add association on GroupUser

robot269006 committed:
Revision (enable instance for redirect_to)

1. Add (@group) to redirect_to root_path, since redirect_to method does
not pass on the instance information.

robot269006 committed:
Revision (edit render)

render action: :new -> render :new
This makes it not go through the controller

ryouzi commented on db/schema.rb:
```diff
@@ -10,12 +10,21 @@
+ActiveRecord::Schema.define(version: 20170419112916) do
+
+  create_table "group_users", force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8" do |t|
+    t.integer  "users_id"
```
命名規則についてなのですが、
user_id, group_id
ではないでしょうか？

ryouzi commented on app/models/user.rb:
```diff
@@ -1,6 +1,8 @@
-  # :confirmable, :lockable, :timeoutable and :omniauthable
+  has_many :groups, through: :group_user
+  has_many :group_users
+  accepts_nested_attributes_for :groups
```
userモデルにもaccepts_nested_attributes_forは必要でしょうか？

robot269006 committed:
Revision (Rename group_user columns, delete attribute nesting on user.rb)

1. Rename group_user columns
2. Delete attribute nesting on user.rb

ryouzi commented on db/schema.rb:
```diff
@@ -37,4 +46,6 @@
 
+  add_foreign_key "group_users", "groups"
+  add_foreign_key "group_users", "users"
 end
```
さきほど指摘したところの
users_idとgroups_idはどのように変更しましたか？

robot269006 replied:
migrationファイルを作成して、その中に下記を記述してカラム名の修正をしてます。
このままだとforeign_keyがうまく動作しないということですか？
    rename_column :group_users, :users_id, :user_id
    rename_column :group_users, :groups_id, :group_id

ryouzi replied:
なるほどなるほど。
migrationファイルが多かったので、少し混乱していました。
このままで大丈夫です。

ryouzi commented:
LGTM
